File: selective_search/selective_search.cpp
#include "opencv2/ximgproc/segmentation.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/core.hpp"
#include "opencv2/imgproc.hpp"
#include <iostream>

using namespace std;
using namespace cv;

static void help() {
	cout << "Usage: " << endl;
	cout << "./ssearch input image (f|q)" << endl;
	cout << "f = fast, q = quality" << endl;
	cout << "Use l to display less rects..." << endl;
	cout << "...m to display more rects..." << endl;
	cout << "...x to quit." << endl;
}

int main(int argc, char** argv) {
	
	if (argc < 3) {
		help();
		return -1;
	}

	// Speed-up using multi-threads
	setUseOptimized(true);		// enable the optimized code
	setNumThreads(4);

	Mat img = imread(argv[1]);
	
	//resize
	int newHeight = 200;
	int newWidth = img.cols * newHeight / img.rows;
	resize(img, img, Size(newWidth, newHeight));

	// Create Selective Search Segmentation Object
	//                  with defautl parameter 
	Ptr<ximgproc::segmentation::SelectiveSearchSegmentation> ss =
		ximgproc::segmentation::createSelectiveSearchSegmentation();
	ss->setBaseImage(img);

	// "f" mode
	// Switch to fast but low recall Selective Search method
	if (argv[2][0] == 'f') {
		ss->switchToSelectiveSearchFast();
	}
	// "q" mode
	// Switch to high recall but slow Selective Search method
	else if (argv[2][0] == 'q') {
		ss->switchToSelectiveSearchQuality();
	}
	else {
		help();
		return -2;
	}

	// run Selective Search Segmentation on input image
	vector<Rect> rects;
	ss->process(rects);

	cout << "Total Number of Region Proposals : " <<
		rects.size() << endl;

	// number of region proposals to show
	int numShowRects = 100;
	// increment to increase/decrease total number of 
	// region proposals to be shown
	int increment = 50;

	while (1) {
		// create a copy of original image
		Mat dest = img.clone();

		// iterate over all the region proposals
		for (int i = 0; i < rects.size(); i++) {
			if (i < numShowRects) {
				rectangle(dest, rects[i], Scalar(255,0,0));
			}
			else {
				break;
			}
		}

		// show output
		imshow("SELECTVIE SEARCH RESULT", dest);

		// record the key press
		int k = waitKey();

		// "m (more) " is pressed
		if (k == 'm') {
			// increase total number of rectangles to show 
			numShowRects += increment;
		}
		// "l (less)" is pressed
		else if (k == 'l' && numShowRects > increment) {
			// decrease total number of rectangles
			numShowRects -= increment;
		}
		// "q" is pressed
		else if (k == 'x') {
			break;
		}
	}
	return 0;
}

File: selective_search/selective_search.py
#!/usr/bin/env python
'''
Usage:
    ./selctive_search.py input_image (f|q)
    f = fast, q = quality
Use 'l' to display less rects, 'm' to display more rects,
'x' to exit
'''
import sys
import cv2

if __name__ == '__main__':
    if len(sys.argv) < 3 :
        print(__doc__)
        sys.exit(1)

    # Speed-up using multi-threads
    cv2.setUseOptimized(True)
    cv2.setNumThreads(4)

    img = cv2.imread(sys.argv[1])
    # resize
    newHeight = 200
    newWidth = int(img.shape[1] * newHeight / img.shape[0])
    img = cv2.resize(img, (newWidth, newHeight))

    # create Selective search Segmentation Object
    #                   with default parameter
    ss = cv2.ximgproc.segmentation.createSelectiveSearchSegmentation()
    ss.setBaseImage(img)

    # "f" mode
    # Switch to fast but low recall Selective Search method
    if ( sys.argv[2] == 'f'):
        ss.switchToSelectiveSearchFast()

    # "q" mode
    # Switch to high recall but slow Selective Search method
    elif ( sys.argv[2] == 'q'):
        ss.switchToSelectiveSearchQuality()

    else:
        print(__doc__)
        sys.exit(1)

    # run Selective Search segmentation on input image
    rects = ss.process()
    print('Total Number of Region Proposals:{}'.format(len(rects)))

    # number of region proposals to show
    numShowRects = 100
    # increment to increase/decrease total number of region proposals to be shown
    increment = 50

    while True:
        dest = img.copy()

        # iterate over all the region proposals
        for i, rect in enumerate(rects):
            if (i < numShowRects):
                x, y, w, h = rect
                cv2.rectangle(dest, (x,y), (x+w, y+h), (255,0,0), 1, cv2.LINE_AA)
            else:
                break
        
        # show output
        cv2.imshow("SELECTIVE SEARCH RESULT", dest)

        # record the key press
        k = cv2.waitKey(0) & 0xFF

        # 'm(more)' is pressed
        if  k == ord('m'):
            numShowRects += increment

        # 'l(less)' is pressed
        elif k == ord('l') and numShowRects > increment:
            numShowRects -= increment

        # 'x' is pressed
        elif k == ord('x'):
            break

    # close image show window
    cv2.destroyAllWindows()

